File: serveur/routes/parametres.py
import json
import os
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import Optional, Literal
from serveur.services.service_snmp import tester_connexion_snmp

router = APIRouter()

CONFIG_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "config")
CONFIG_FILE = os.path.join(CONFIG_DIR, "settings.json")

# Configuration par défaut
DEFAULT_SETTINGS = {
    "mode": "MOCK",  # "MOCK" ou "REAL_NETWORK"
    "core_switch_ip": "10.20.0.1",
    "core_switch_brand": "Cisco",
    "snmp_version": "v2c",
    "snmp_community": "ASECNA_READ",
    "snmp_v3_user": "asecna_admin",
    "snmp_v3_auth_key": "AuthKeySecure123",
    "snmp_v3_priv_key": "PrivKeySecure123",
    "snmp_v3_auth_proto": "SHA",
    "snmp_v3_priv_proto": "AES",
    "ssh_username": "admin",
    "ssh_password": "adminpassword",
    "theme": "Dark NOC",
    "custom_logo_path": "../image/Logo_ASECNA.png"
}

def charger_parametres():
    if not os.path.exists(CONFIG_DIR):
        os.makedirs(CONFIG_DIR, exist_ok=True)
    if not os.path.exists(CONFIG_FILE):
        with open(CONFIG_FILE, "w", encoding="utf-8") as f:
            json.dump(DEFAULT_SETTINGS, f, indent=4)
        return DEFAULT_SETTINGS
    try:
        with open(CONFIG_FILE, "r", encoding="utf-8") as f:
            return json.load(f)
    except Exception:
        return DEFAULT_SETTINGS

def sauvegarder_parametres(data: dict):
    if not os.path.exists(CONFIG_DIR):
        os.makedirs(CONFIG_DIR, exist_ok=True)
    with open(CONFIG_FILE, "w", encoding="utf-8") as f:
        json.dump(data, f, indent=4)

class SchemaParametres(BaseModel):
    mode: Literal["MOCK", "REAL_NETWORK"] = "MOCK"
    core_switch_ip: str
    core_switch_brand: Literal["Cisco", "Huawei", "Generic"] = "Cisco"
    snmp_version: Literal["v2c", "v3"] = "v2c"
    snmp_community: Optional[str] = "public"
    snmp_v3_user: Optional[str] = ""
    snmp_v3_auth_key: Optional[str] = ""
    snmp_v3_priv_key: Optional[str] = ""
    snmp_v3_auth_proto: Optional[Literal["MD5", "SHA"]] = "SHA"
    snmp_v3_priv_proto: Optional[Literal["DES", "AES"]] = "AES"
    ssh_username: Optional[str] = ""
    ssh_password: Optional[str] = ""
    theme: Optional[Literal["Dark NOC", "Cyberpunk", "Light Professional"]] = "Dark NOC"
    custom_logo_path: Optional[str] = "../image/Logo_ASECNA.png"

class RequeteTestSNMP(BaseModel):
    ip: str
    version: str
    community: Optional[str] = "public"
    v3_user: Optional[str] = ""
    v3_auth_key: Optional[str] = ""
    v3_priv_key: Optional[str] = ""
    v3_auth_proto: Optional[str] = "SHA"
    v3_priv_proto: Optional[str] = "AES"

@router.get("", response_model=SchemaParametres)
async def obtenir_parametres():
    """Récupère les paramètres de configuration actuels."""
    return charger_parametres()

@router.post("")
async def enregistrer_parametres(params: SchemaParametres):
    """Enregistre les nouveaux paramètres (Réservé ADMIN)."""
    sauvegarder_parametres(params.dict())
    return {"succes": True, "message": "Paramètres réseau mis à jour avec succès."}

@router.post("/tester")
async def tester_snmp(requete: RequeteTestSNMP):
    """Teste la connexion SNMP vers l'IP Cœur configurée."""
    resultat = tester_connexion_snmp(
        ip=requete.ip,
        version=requete.version,
        community=requete.community,
        v3_user=requete.v3_user,
        v3_auth_key=requete.v3_auth_key,
        v3_priv_key=requete.v3_priv_key,
        v3_auth_proto=requete.v3_auth_proto,
        v3_priv_proto=requete.v3_priv_proto
    )
    if not resultat.get("succes"):
        raise HTTPException(status_code=400, detail=resultat.get("message", "Échec du test SNMP."))
    return resultat

File: interface/ressources/js/parametres.js
/**
 * parametres.js — Gestion de la configuration réseau dynamique (Live & Mock)
 */

document.addEventListener('DOMContentLoaded', () => {
    const form = document.getElementById('form-settings');
    const selectVersion = document.getElementById('setting-snmp-version');
    const containerV2 = document.getElementById('snmp-v2c-fields');
    const containerV3 = document.getElementById('snmp-v3-fields');
    const btnTest = document.getElementById('btn-test-snmp');
    const testResult = document.getElementById('snmp-test-result');

    // 1. Bascule dynamique SNMP v2c / v3
    selectVersion?.addEventListener('change', () => {
        if (selectVersion.value === 'v2c') {
            containerV2?.classList.remove('hidden');
            containerV3?.classList.add('hidden');
        } else {
            containerV2?.classList.add('hidden');
            containerV3?.classList.remove('hidden');
        }
    });

    // 2. Charger les paramètres existants depuis l'API
    chargerParametres();

    // 3. Tester la connexion SNMP
    btnTest?.addEventListener('click', async () => {
        const ip = document.getElementById('setting-core-ip').value.trim();
        const version = selectVersion.value;
        const community = document.getElementById('setting-snmp-community').value.trim();

        if (!ip) {
            alert("Veuillez renseigner l'adresse IP du switch cœur.");
            return;
        }

        const originalText = btnTest.innerHTML;
        btnTest.disabled = true;
        btnTest.innerHTML = '<i class="fa-solid fa-spinner fa-spin mr-1"></i> Test en cours...';
        testResult?.classList.add('hidden');

        const payload = {
            ip: ip,
            version: version,
            community: community,
            v3_user: document.getElementById('setting-snmp-v3-user')?.value || "",
            v3_auth_key: document.getElementById('setting-snmp-v3-auth-key')?.value || "",
            v3_priv_key: document.getElementById('setting-snmp-v3-priv-key')?.value || "",
            v3_auth_proto: document.getElementById('setting-snmp-v3-auth-proto')?.value || "SHA",
            v3_priv_proto: document.getElementById('setting-snmp-v3-priv-proto')?.value || "AES"
        };

        try {
            const resp = await fetch('http://localhost:8000/api/v1/parametres/tester', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(payload)
            });
            const data = await resp.json();

            testResult.classList.remove('hidden');
            if (resp.ok && data.succes) {
                testResult.className = "p-3 rounded-lg text-xs bg-emerald-500/10 border border-emerald-500/30 text-emerald-400";
                testResult.innerHTML = `
                    <div class="flex items-center gap-2 font-bold mb-1">
                        <i class="fa-solid fa-circle-check"></i> ${data.message}
                    </div>
                    <div class="font-mono text-[11px] text-gray-300">
                        Nœud: <strong>${data.sysName}</strong> | Latence: ${data.latence_ms}ms<br>
                        Descr: ${data.sysDescr}
                    </div>
                `;
            } else {
                testResult.className = "p-3 rounded-lg text-xs bg-danger/10 border border-danger/30 text-danger";
                testResult.innerHTML = `<i class="fa-solid fa-circle-xmark mr-1"></i> ${data.detail || data.message || "Échec de connexion SNMP"}`;
            }
        } catch (e) {
            testResult.classList.remove('hidden');
            testResult.className = "p-3 rounded-lg text-xs bg-danger/10 border border-danger/30 text-danger";
            testResult.innerHTML = `<i class="fa-solid fa-triangle-exclamation mr-1"></i> Erreur lors de l'appel au serveur API.`;
        } finally {
            btnTest.disabled = false;
            btnTest.innerHTML = originalText;
        }
    });

    // 4. Enregistrer les paramètres
    form?.addEventListener('submit', async (e) => {
        e.preventDefault();
        const submitBtn = document.getElementById('btn-save-settings');
        const origBtnText = submitBtn.innerHTML;
        submitBtn.disabled = true;
        submitBtn.innerHTML = '<i class="fa-solid fa-spinner fa-spin mr-1"></i> Enregistrement...';

        const payload = {
            mode: document.getElementById('setting-mode').value,
            core_switch_ip: document.getElementById('setting-core-ip').value.trim(),
            core_switch_brand: document.getElementById('setting-core-brand').value,
            snmp_version: selectVersion.value,
            snmp_community: document.getElementById('setting-snmp-community').value.trim(),
            snmp_v3_user: document.getElementById('setting-snmp-v3-user').value.trim(),
            snmp_v3_auth_key: document.getElementById('setting-snmp-v3-auth-key').value,
            snmp_v3_priv_key: document.getElementById('setting-snmp-v3-priv-key').value,
            snmp_v3_auth_proto: document.getElementById('setting-snmp-v3-auth-proto').value,
            snmp_v3_priv_proto: document.getElementById('setting-snmp-v3-priv-proto').value,
            ssh_username: document.getElementById('setting-ssh-user').value.trim(),
            ssh_password: document.getElementById('setting-ssh-pwd').value,
            theme: document.getElementById('setting-theme').value,
            custom_logo_path: document.getElementById('setting-logo-path').value.trim()
        };

        try {
            const resp = await fetch('http://localhost:8000/api/v1/parametres', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify(payload)
            });
            if (resp.ok) {
                afficherToastParam("Paramètres réseau et SNMP enregistrés avec succès !", "success");
            } else {
                afficherToastParam("Erreur lors de l'enregistrement.", "error");
            }
        } catch (e) {
            afficherToastParam("Impossible de joindre le serveur API.", "error");
        } finally {
            submitBtn.disabled = false;
            submitBtn.innerHTML = origBtnText;
        }
    });
});

async function chargerParametres() {
    try {
        const resp = await fetch('http://localhost:8000/api/v1/parametres');
        if (!resp.ok) return;
        const data = await resp.json();

        if (document.getElementById('setting-mode')) document.getElementById('setting-mode').value = data.mode || "MOCK";
        if (document.getElementById('setting-core-ip')) document.getElementById('setting-core-ip').value = data.core_switch_ip || "10.20.0.1";
        if (document.getElementById('setting-core-brand')) document.getElementById('setting-core-brand').value = data.core_switch_brand || "Cisco";
        
        const selVer = document.getElementById('setting-snmp-version');
        if (selVer) {
            selVer.value = data.snmp_version || "v2c";
            selVer.dispatchEvent(new Event('change'));
        }

        if (document.getElementById('setting-snmp-community')) document.getElementById('setting-snmp-community').value = data.snmp_community || "public";
        if (document.getElementById('setting-snmp-v3-user')) document.getElementById('setting-snmp-v3-user').value = data.snmp_v3_user || "";
        if (document.getElementById('setting-snmp-v3-auth-key')) document.getElementById('setting-snmp-v3-auth-key').value = data.snmp_v3_auth_key || "";
        if (document.getElementById('setting-snmp-v3-priv-key')) document.getElementById('setting-snmp-v3-priv-key').value = data.snmp_v3_priv_key || "";
        if (document.getElementById('setting-snmp-v3-auth-proto')) document.getElementById('setting-snmp-v3-auth-proto').value = data.snmp_v3_auth_proto || "SHA";
        if (document.getElementById('setting-snmp-v3-priv-proto')) document.getElementById('setting-snmp-v3-priv-proto').value = data.snmp_v3_priv_proto || "AES";
        if (document.getElementById('setting-ssh-user')) document.getElementById('setting-ssh-user').value = data.ssh_username || "";
        if (document.getElementById('setting-ssh-pwd')) document.getElementById('setting-ssh-pwd').value = data.ssh_password || "";
        if (document.getElementById('setting-theme')) document.getElementById('setting-theme').value = data.theme || "Dark NOC";
        if (document.getElementById('setting-logo-path')) document.getElementById('setting-logo-path').value = data.custom_logo_path || "../image/Logo_ASECNA.png";

    } catch (e) {
        console.warn("Impossible de charger les paramètres réseau depuis l'API.");
    }
}

function afficherToastParam(message, type = 'success') {
    const toast = document.createElement('div');
    const ok = type === 'success';
    toast.style.cssText = `
        position:fixed;bottom:24px;right:24px;z-index:9999;
        padding:12px 18px;border-radius:10px;font-size:13px;font-weight:500;
        display:flex;align-items:center;gap:10px;
        background:${ok ? 'rgba(16,185,129,0.15)' : 'rgba(239,68,68,0.15)'};
        border:1px solid ${ok ? 'rgba(16,185,129,0.4)' : 'rgba(239,68,68,0.4)'};
        color:${ok ? '#10b981' : '#f87171'};
        box-shadow:0 8px 24px rgba(0,0,0,0.5);
    `;
    toast.innerHTML = `<i class="fa-solid fa-${ok ? 'circle-check' : 'triangle-exclamation'}"></i> ${message}`;
    document.body.appendChild(toast);
    setTimeout(() => toast.remove(), 4000);
}